pkg/status/singletonstatus.go

```diff
@@ -18,69 +18,78 @@ package status
+	for i := range wObjIDs {
+		wObjID, r, n := wObjIDs[i], requested[i], nWECs[i]
+		if !r || n != 1 {
+			if err := c.reconcileSingletonWObj(ctx, wObjID, false); err != nil {
+				return err
+			}
+			logger.V(4).Info("Cleaned singleton status for workload object",
+				"gvk", wObjID.GVK, "objectName", wObjID.ObjectName,
+				"requested", r, "nWECs", n)
+		} else {
+			if err := c.reconcileSingletonWObj(ctx, wObjID, true); err != nil {
+				return err
+			}
+			logger.V(4).Info("Updated singleton status for workload object",
+				"gvk", wObjID.GVK, "objectName", wObjID.ObjectName)
 		}
-		return fmt.Errorf("could not get binding %s: %w", bdgName, err)
 	}
+	return nil
+}
 
-	sync := make(map[util.ObjectIdentifier]bool)
-	if err = c.checkSingletonWorkloadInBinding(ctx, *binding, sync, true); err != nil {
+func (c *Controller) reconcileSingletonByWS(ctx context.Context, ref singletonWorkStatusRef) error {
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Reconciling singleton status due to workstatus changes", "name", string(ref.Name))
+
+	wObjID := ref.SourceObjectIdentifier
+	requested, nWECs := c.bindingPolicyResolver.GetSingletonReportedStateRequestForObject(wObjID)
+	if !requested || nWECs != 1 {
+		if err := c.reconcileSingletonWObj(ctx, wObjID, false); err != nil {
+			return err
+		}
+		logger.V(4).Info("Cleaned singleton status for workload object",
+			"gvk", ref.SourceObjectIdentifier.GVK, "objectName", ref.SourceObjectIdentifier.ObjectName,
+			"requested", requested, "nWECs", nWECs)
+		return nil
+	}
+	wsObj, err := c.workStatusLister.ByNamespace(ref.WECName).Get(ref.Name)
+	if err != nil {
 		return err
 	}
-
-	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
+	status, err := util.GetWorkStatusStatus(wsObj)
 	if err != nil {
 		return err
 	}
-
-	for _, bdg := range allBdgs.Items {
-		if bdg.Name == binding.Name {
-			continue
-		}
-		if err = c.checkSingletonWorkloadInBinding(ctx, bdg, sync, false); err != nil {
-			return err
-		}
+	if status == nil {
+		return nil
 	}
-
-	for wObjIdentifier, v := range sync {
-		if err := c.reconcileSingletonWObj(ctx, wObjIdentifier, v); err != nil {
-			return err
-		}
+	if err := updateObjectStatus(ctx, &wObjID, status, c.listers, c.wdsDynClient); err != nil {
+		return err
 	}
+	logger.V(4).Info("Updated singleton status for workload object",
+		"gvk", ref.SourceObjectIdentifier.GVK, "objectName", ref.SourceObjectIdentifier.ObjectName)
 	return nil
 }
 
 func (c *Controller) reconcileSingletonWObj(ctx context.Context, wObjID util.ObjectIdentifier, sync bool) error {
 	logger := klog.FromContext(ctx)
-	logger.V(4).Info("Reconciling singleton workload object", "resource", wObjID.Resource, "objectName", wObjID.ObjectName)
+	logger.V(4).Info("Reconciling workload object for singleton status", "gvk", wObjID.GVK, "objectName", wObjID.ObjectName, "sync", sync)
```

Comment:
I would make this `V(5)`, since all the call sites also have a `V(4)` reporting the same information.